node/engine/chainservice/eth_chainservice.go

```diff
@@ -356,34 +363,55 @@ func (ecs *EthChainService) dispatchChainEvents(logs []ethTypes.Log) error {
 	return nil
 }
 
-func (ecs *EthChainService) listenForEventLogs(errorChan chan<- error, eventSub ethereum.Subscription, eventChan chan ethTypes.Log, eventQuery ethereum.FilterQuery) {
-out:
+func (ecs *EthChainService) listenForEventLogs(errorChan chan<- error, eventChan chan ethTypes.Log, eventQuery ethereum.FilterQuery) {
 	for {
 		select {
 		case <-ecs.ctx.Done():
-			eventSub.Unsubscribe()
+			ecs.eventSub.Unsubscribe()
 			ecs.wg.Done()
 			return
 
-		case err := <-eventSub.Err():
+		case err := <-ecs.eventSub.Err():
+			ecs.eventTracker.mu.Lock()
+			defer ecs.eventTracker.mu.Unlock()
+
+			latestBlockNum := ecs.eventTracker.latestBlockNum
+
 			if err != nil {
 				ecs.logger.Warn("error in chain event subscription: " + err.Error())
-				eventSub.Unsubscribe()
+				ecs.eventSub.Unsubscribe()
+			} else {
+				ecs.logger.Warn("chain event subscription closed")
 			}
 
-			// Try to re-establish subscription once before failing
-			eventSub, err = ecs.chain.SubscribeFilterLogs(ecs.ctx, eventQuery, eventChan)
-			if err != nil {
-				errorChan <- fmt.Errorf("subscribeFilterLogs failed to resubscribe: " + err.Error())
-				break out
+			// Use exponential backoff loop to attempt to re-establish subscription
+			for backoffTime := MIN_BACKOFF_TIME; backoffTime < MAX_BACKOFF_TIME; backoffTime *= 2 {
```

Comment:
Instead of exponential backoff should this attempt the re-connect at a consistent interval?

Reply:
Seems marginal to me. I think [exponential backoff](https://en.wikipedia.org/wiki/Exponential_backoff) is fine, although I'm not sure if it is giving us much benefit over a consistent interval. It might help us avoid being rate limited by the chain endpoint.